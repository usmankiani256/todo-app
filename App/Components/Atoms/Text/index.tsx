import React from 'react'
import {
  Text as RNText,
  TextStyle,
  TextProps as RNTextProps,
} from 'react-native'
import { scaleFont } from '@Utils'
import { Colors, FontFamily } from '@Theme'

interface TextProps extends RNTextProps {
  children: React.ReactNode
  h1?: boolean
  h2?: boolean
  h3?: boolean
  h4?: boolean
  h5?: boolean
  w1?: boolean
  w2?: boolean
  w3?: boolean
  w4?: boolean
  w5?: boolean
  p?: boolean
  bold?: boolean
  italic?: boolean
  center?: boolean
  justify?: boolean
  selfCenter?: boolean
  style?: TextStyle
}

const Text = (props: TextProps) => {
  const {
    children,
    h1,
    h2,
    h3,
    h4,
    h5 = true,
    w1,
    w2,
    w3,
    w4 = true,
    w5,
    p,
    bold,
    italic,
    center,
    justify,
    selfCenter,
    style,
    ...rest
  } = props

  let textStyle: TextStyle = {
    color: Colors.text,
  }

  if (h1) {
    textStyle.fontSize = scaleFont(48)
    textStyle.lineHeight = scaleFont(48)
  } else if (h2) {
    textStyle.fontSize = scaleFont(34)
    textStyle.lineHeight = scaleFont(34)
  } else if (h3) {
    textStyle.fontSize = scaleFont(24)
    textStyle.lineHeight = scaleFont(24)
  } else if (h4) {
    textStyle.fontSize = scaleFont(20)
    textStyle.lineHeight = scaleFont(20)
  } else if (p) {
    textStyle.fontSize = scaleFont(14)
    textStyle.lineHeight = scaleFont(14)
  } else if (h5) {
    textStyle.fontSize = scaleFont(16)
    textStyle.lineHeight = scaleFont(16)
  }

  if (w1) textStyle.fontFamily = FontFamily.Bold
  else if (w2) textStyle.fontFamily = FontFamily.SemiBold
  else if (w3) textStyle.fontFamily = FontFamily.Medium
  else if (w5) textStyle.fontFamily = FontFamily.Light
  else if (w4) textStyle.fontFamily = FontFamily.Regular

  if (bold) textStyle.fontWeight = 'bold'
  if (italic) textStyle.fontStyle = 'italic'
  if (center) textStyle.textAlign = 'center'
  if (justify) textStyle.textAlign = 'justify'
  if (selfCenter) textStyle.alignSelf = 'center'

  if (style) textStyle = { ...textStyle, ...style }

  return (
    <RNText style={textStyle} {...rest}>
      {children}
    </RNText>
  )
}

export default Text
