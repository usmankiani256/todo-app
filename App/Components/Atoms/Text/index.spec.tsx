import * as React from 'react'
import { render } from '@testing-library/react-native'
import Text from './index'

// Arrange
const children = 'Hello world!'

describe('Text Component', () => {
  it('renders text correctly', () => {
    // Act
    const rendered: any = render(<Text>{children}</Text>).toJSON()

    // Assert
    expect(rendered).toMatchSnapshot()
  })

  it('renders props', () => {
    // Arrange
    const numberOfLines = 4
    const customStyle = {
      padding: 10,
    }

    // Act
    const rendered: any = render(
      <Text
        h1
        h2
        bold
        italic
        center
        justify
        selfCenter
        style={customStyle}
        numberOfLines={numberOfLines}>
        {children}
      </Text>,
    ).toJSON()

    // Assert
    expect(rendered).toMatchSnapshot()
  })

  it('renders h2 and w2', () => {
    // Act
    const rendered: any = render(
      <Text h2 w2>
        {children}
      </Text>,
    ).toJSON()

    // Assert
    expect(rendered).toMatchSnapshot()
  })

  it('renders h3 and w3', () => {
    // Act
    const rendered: any = render(
      <Text h3 w3>
        {children}
      </Text>,
    ).toJSON()

    // Assert
    expect(rendered).toMatchSnapshot()
  })

  it('renders h4 and w4', () => {
    // Act
    const rendered: any = render(
      <Text h4 w4>
        {children}
      </Text>,
    ).toJSON()

    // Assert
    expect(rendered).toMatchSnapshot()
  })

  it('renders h5 and w5', () => {
    // Act
    const rendered: any = render(
      <Text h5 w5>
        {children}
      </Text>,
    ).toJSON()

    // Assert
    expect(rendered).toMatchSnapshot()
  })

  it('renders p', () => {
    // Act
    const rendered: any = render(<Text p>{children}</Text>).toJSON()

    // Assert
    expect(rendered).toMatchSnapshot()
  })
})
