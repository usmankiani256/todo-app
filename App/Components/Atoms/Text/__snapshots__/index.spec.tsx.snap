// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component renders h2 and w2 1`] = `
<Text
  style={
    Object {
      "color": "#3f3f3f",
      "fontFamily": "SofiaPro-SemiBold",
      "fontSize": 42.5,
      "lineHeight": 42.5,
    }
  }
>
  Hello world!
</Text>
`;

exports[`Text Component renders h3 and w3 1`] = `
<Text
  style={
    Object {
      "color": "#3f3f3f",
      "fontFamily": "SofiaPro-Medium",
      "fontSize": 30,
      "lineHeight": 30,
    }
  }
>
  Hello world!
</Text>
`;

exports[`Text Component renders h4 and w4 1`] = `
<Text
  style={
    Object {
      "color": "#3f3f3f",
      "fontFamily": "SofiaPro",
      "fontSize": 25,
      "lineHeight": 25,
    }
  }
>
  Hello world!
</Text>
`;

exports[`Text Component renders h5 and w5 1`] = `
<Text
  style={
    Object {
      "color": "#3f3f3f",
      "fontFamily": "SofiaPro-Light",
      "fontSize": 20,
      "lineHeight": 20,
    }
  }
>
  Hello world!
</Text>
`;

exports[`Text Component renders p 1`] = `
<Text
  style={
    Object {
      "color": "#3f3f3f",
      "fontFamily": "SofiaPro",
      "fontSize": 17.5,
      "lineHeight": 17.5,
    }
  }
>
  Hello world!
</Text>
`;

exports[`Text Component renders props 1`] = `
<Text
  numberOfLines={4}
  style={
    Object {
      "alignSelf": "center",
      "color": "#3f3f3f",
      "fontFamily": "SofiaPro",
      "fontSize": 60,
      "fontStyle": "italic",
      "fontWeight": "bold",
      "lineHeight": 60,
      "padding": 10,
      "textAlign": "justify",
    }
  }
>
  Hello world!
</Text>
`;

exports[`Text Component renders text correctly 1`] = `
<Text
  style={
    Object {
      "color": "#3f3f3f",
      "fontFamily": "SofiaPro",
      "fontSize": 20,
      "lineHeight": 20,
    }
  }
>
  Hello world!
</Text>
`;
